onnxruntime/core/framework/utils.h

```diff
@@ -87,69 +87,254 @@ void DumpNodeInputs(const OpKernelContext& context, const Node& node);
 void DumpNodeOutputs(OpKernelContext& context, const Node& node, const SessionState& session_state);
 #endif
 
-#define DispatchOnTensorType(tensor_type, function, ...)        \
-  if (tensor_type == DataTypeImpl::GetType<float>())            \
-    function<float>(__VA_ARGS__);                               \
-  else if (tensor_type == DataTypeImpl::GetType<double>())      \
-    function<double>(__VA_ARGS__);                              \
-  else if (tensor_type == DataTypeImpl::GetType<int8_t>())      \
-    function<int8_t>(__VA_ARGS__);                              \
-  else if (tensor_type == DataTypeImpl::GetType<int16_t>())     \
-    function<int16_t>(__VA_ARGS__);                             \
-  else if (tensor_type == DataTypeImpl::GetType<int32_t>())     \
-    function<int32_t>(__VA_ARGS__);                             \
-  else if (tensor_type == DataTypeImpl::GetType<int64_t>())     \
-    function<int64_t>(__VA_ARGS__);                             \
-  else if (tensor_type == DataTypeImpl::GetType<uint8_t>())     \
-    function<uint8_t>(__VA_ARGS__);                             \
-  else if (tensor_type == DataTypeImpl::GetType<uint16_t>())    \
-    function<uint16_t>(__VA_ARGS__);                            \
-  else if (tensor_type == DataTypeImpl::GetType<uint32_t>())    \
-    function<uint32_t>(__VA_ARGS__);                            \
-  else if (tensor_type == DataTypeImpl::GetType<uint64_t>())    \
-    function<uint64_t>(__VA_ARGS__);                            \
-  else if (tensor_type == DataTypeImpl::GetType<bool>())        \
-    function<bool>(__VA_ARGS__);                                \
-  else if (tensor_type == DataTypeImpl::GetType<MLFloat16>())   \
-    function<MLFloat16>(__VA_ARGS__);                           \
-  else if (tensor_type == DataTypeImpl::GetType<BFloat16>())    \
-    function<BFloat16>(__VA_ARGS__);                            \
-  else if (tensor_type == DataTypeImpl::GetType<std::string>()) \
-    function<std::string>(__VA_ARGS__);                         \
-  else                                                          \
-    ORT_ENFORCE(false, "Unknown tensor type of ", tensor_type)
+template <typename T>
+constexpr ONNX_NAMESPACE::TensorProto_DataType ToTensorDataType();
```

Comment:
Should we split these out into something like data_types_internal.h? They don't feel like they fit in a general header of utility methods.

Reply:
Still wondering about this. There seems to be more than enough stuff to have something like data_type_utils.h with the data type related stuff.

Reply:
Done.

---
In reply to: [342908804](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r342908804) [](ancestors = 342908804)